src/i256.rs

```diff
@@ -0,0 +1,279 @@
+// UInt words count = 256 / 64 = 4
+construct_uint! {
+    pub struct U256(4);
+}
+
+/// Signed 256-bit number. Works on top of U256 with help of two's complement.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct I256 {
+    inner: U256,
+}
+
+impl I256 {
+    /// Value `i128::MAX`
+    pub const I128_MAX: Self = Self::from_i128(i128::MAX);
+    /// Value `i128::MIN`. Very useful because `abs` isn't defined for `i128::MIN`
+    pub const I128_MIN: Self = Self::from_i128(i128::MIN);
+    pub const MAX: Self = Self::new(U256([u64::MAX, u64::MAX, u64::MAX, !SIGN_MASK]));
+    pub const MIN: Self = Self::new(U256([0, 0, 0, SIGN_MASK]));
+
+    const fn new(x: U256) -> Self {
+        I256 { inner: x }
+    }
+
+    pub const fn from_i128(x: i128) -> Self {
+        let msb = if x < 0 { u64::MAX } else { 0 };
+        Self::new(U256([
+            x as u64,
+            (x >> UINT_WORD_BITS_COUNT) as u64,
+            msb,
+            msb,
+        ])) // The only way to do it const
+    }
+
+    fn abs(self) -> Self {
+        if !self.is_negative() {
+            // positive or zero
+            return self;
+        }
+        -self
+    }
+
+    const fn is_negative(self) -> bool {
+        self.sign() != 0
+    }
+
+    /// 63'rd bit shows number sign:
+    /// 1 -- number < 0 (0x8000_0000_0000_0000),
+    /// 0 -- number >= 0 (0).
+    /// Other bits are equal to 0.
+    const fn sign(self) -> u64 {
+        let most_significant_word: u64 = self.words()[UINT_WORDS_COUNT - 1];
+        most_significant_word & SIGN_MASK
+    }
+
+    const fn words<'a>(&'a self) -> &'a [u64; UINT_WORDS_COUNT] {
+        &self.inner.0
+    }
+}
+
+impl Mul for I256 {
+    type Output = Self;
+
+    fn mul(self, rhs: Self) -> Self::Output {
+        let lhs_sign = self.sign();
+        let rhs_sign = rhs.sign();
+
+        let lhs = if lhs_sign == 0 { self } else { -self };
+        let rhs = if rhs_sign == 0 { rhs } else { -rhs };
+
+        // Mustn't overflow because we're usually promoting just i128 to I256.
+        let result = Self::new(lhs.inner * rhs.inner);
+        if lhs_sign ^ rhs_sign == 0 {
+            result
+        } else {
+            -result
+        }
+    }
+}
+
+impl Div for I256 {
+    type Output = Self;
+
+    fn div(self, rhs: Self) -> Self::Output {
+        let lhs_sign = self.sign();
+        let rhs_sign = rhs.sign();
+
+        let lhs = if lhs_sign == 0 { self } else { -self };
+        let rhs = if rhs_sign == 0 { rhs } else { -rhs };
+
+        // Mustn't overflow because we're usually promoting just i128 to I256.
+        let result = Self::new(lhs.inner / rhs.inner);
+        if lhs_sign ^ rhs_sign == 0 {
+            result
+        } else {
+            -result
+        }
+    }
+}
+
+impl Sub for I256 {
+    type Output = Self;
+
+    fn sub(self, rhs: Self) -> Self::Output {
+        let lhs_sign = self.sign();
+        let rhs_sign = rhs.sign();
+
+        let lhs = if lhs_sign == 0 { self } else { -self };
```

Comment:
Probably calls of `abs()` will be optimized anyway.

Reply:
I've decided to remove `abs()` :)